client/src/main/kotlin/io/spine/examples/pingh/client/DesktopClient.kt

```diff
@@ -102,147 +62,57 @@ public class DesktopClient(
-            .select(UserMentions::class.java)
-            .byId(UserMentionsId::class.buildBy(session!!.username))
-            .run()
-        if (userMentions.size == 0) {
-            return listOf()
-        }
-        return userMentions[0]
-            .mentionList
-            .sortedByDescending { mention -> mention.whenMentioned.seconds }
-    }
-
-    /**
-     * Marks the mention as snoozed.
-     *
-     * If snooze time is not specified, the mention will snooze the [defaultSnoozeTime].
-     */
-    public fun markMentionAsSnoozed(
-        id: MentionId,
-        snoozeTime: Duration = defaultSnoozeTime,
-        onSuccess: (event: MentionSnoozed) -> Unit = {}
-    ) {
-        val command = SnoozeMention::class.buildBy(id, currentTime().add(snoozeTime))
-        observeEventOnce(command.id, MentionSnoozed::class, onSuccess)
-        send(command)
-    }
-
-    /**
-     * Marks that the mention is read.
-     */
-    public fun markMentionAsRead(id: MentionId, onSuccess: (event: MentionRead) -> Unit = {}) {
-        val command = MarkMentionAsRead::class.buildBy(id)
-        observeEventOnce(command.id, MentionRead::class, onSuccess)
-        send(command)
     }
 
     /**
      * Sends a command to the server on behalf of the user.
      */
-    private fun send(command: CommandMessage) {
+    internal fun send(command: CommandMessage) {
         clientRequest()
             .command(command)
             .postAndForget()
     }
 
     /**
-     * Provides `ClientRequest` on behalf of logged-in user if it exists,
-     * or as guest if it doesn't.
+     * Reads the entity state by the specified ID.
+     *
+     * @param id the ID of the entity state.
+     * @param type the type of the entity state.
+     * @return the entity state if it exists, or `null` otherwise.
      */
-    private fun clientRequest(): ClientRequest {
-        if (session == null) {
-            return client.asGuest()
-        }
-        return client.onBehalfOf(user)
-    }
+    internal fun <I : Message, E : EntityState> readById(id: I, type: KClass<E>): E? =
+        clientRequest()
+            .select(type.java)
+            .byId(id)
+            .run()
+            .getOrNull(0)
 
     /**
-     * Observes the outcome of the command.
+     * Observes both events until one is emitted.
      *
-     * When a success or failure event is emitted, subscriptions are cancelled.
+     * When either the first or second event is emitted, all subscriptions are cancelled.
+     *
+     * @param id the ID of the events to be observed.
+     * @param firstType the type of the first observed event.
+     * @param onFirst called when the first event is emitted.
+     * @param secondType the type of the second observed event.
+     * @param onSecond called when the second event is emitted.
      */
-    private fun <S : EventMessage, F : EventMessage> observeCommandOutcome(
+    internal fun <F : EventMessage, S : EventMessage> observeEither(
```

Comment:
How about this:

```kotlin
internal data class EventObserver<E: Any> internal constructor(
    internal val id: Message,
    internal val type: KClass<E>,
    internal val callback: (event: E) -> Unit)

// ...

internal fun <F : EventMessage, S : EventMessage> observeEither(
    first: EventObserver<F>,
    second: EventObserver<S>
) {
    var subscriptionOnSecond: Subscription? = null
    val subscriptionOnFirst = observeEventOnce(first.id, first.type) { event ->
        stopObservation(subscriptionOnSecond!!)
        first.callback(event)
    }
    subscriptionOnSecond = observeEventOnce(second.id, second.type) { event ->
        stopObservation(subscriptionOnFirst)
        second.callback(event)
    }
}
```
I also think that we might even receive a vararg of such observers. And eliminate the code duplication you now have in this `...onFirst`—`onSecond` block.